runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -409,6 +423,41 @@ function LanguageTree:_add_injections()
   return query_time
 end
 
+--- Run an asynchronous parse, calling {on_parse} when complete.
+---
+--- @param range boolean|Range?
+--- @param on_parse fun(trees?: table<integer, TSTree>, err?: string)
+--- @return table<integer, TSTree>? trees the list of parsed trees, if parsing completed synchronously
+function LanguageTree:_async_parse(range, on_parse)
+  local ct = self:_buf().changedtick
+  local total_parse_time = 0
+  local redrawtime = vim.o.redrawtime
+
+  local function step()
+    -- If buffer was changed in the middle of parsing, reset parse state
+    if self:_buf().changedtick ~= ct then
+      self._parser:reset()
```

Comment:
This should also remove the corresponding `self_trees[?]`. Not sure what to do if there are multiple trees, since all of them have to be released, but we can't know which one was released by the parser.

Reply:
Why does tree sitter free the old tree if a parse timed out and the parser was reset? Seems like a bug, given that `ts_tree_free()` also frees that same tree.